fix(history_iterator): messages repeat if before & after not set

EmmmaTech opened the pull request:
## Summary

<!-- What is this pull request for? Does it fix any issues? -->

fixes #1238, an issue introduced with the async iterator rewrite in 3.0 where messages would loop every 100 messages due to the `after` parameter never being updated. I also updated the documentation a bit to clarify this default behavior of the iterator since the original documentation never mentions it.

<!-- Link any issues if applicable.
Use keywords from https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests
-->

<!-- Uncomment based on the type of your changes below -->

## This is a **Code Change**

- [x] I have tested my changes.
- [x] I have updated the documentation to reflect the changes.
- [x] I have run `task pyright` and fixed the relevant issues.

EmmmaTech committed:
fix(history_iterator): messages repeat if before & after not set

teaishealthy reviewed (changes requested):
I can't seem to get more than 100 messages

```py
counter = 0
async for _ in channel.history(limit=200):
    counter += 1

print(counter)  # 100 (?)
```

EmmmaTech committed:
fix: additional creeping issues with the fix in the last commit
in order to fix the functionality, the oldest_first parameter now defaults to True and the after parameter defaults to OLDEST_OBJECT to mimick the desired default behavior of fetching limit messages from the beginning of channel history.

EmmmaTech commented:
this might be more than a fix now, it's a semi-refactor to try and more closely match the original discord.py behavior while fixing the desired result to fetch limit messages from the beginning of channel history. :p

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

teaishealthy commented:
ran a few tests

**master**
```
❌ Failed for params: limit=None, before=None, after=None, around=None, oldest_first=False. Expected 150, got asyncio.TimeoutError.
✅ Passed for params: limit=50, before=None, after=None, around=None, oldest_first=False. Expected 50, got 50.
❌ Failed for params: limit=120, before=None, after=None, around=None, oldest_first=False. Expected 120, got duplicate messages found.
✅ Passed for params: limit=None, before=<message #130>, after=None, around=None, oldest_first=False. Expected 129, got 129.
✅ Passed for params: limit=150, before=<message #130>, after=None, around=None, oldest_first=False. Expected 129, got 129.
✅ Passed for params: limit=80, before=<message #130>, after=None, around=None, oldest_first=False. Expected 80, got 80.
✅ Passed for params: limit=None, before=None, after=<message #20>, around=None, oldest_first=True. Expected 130, got 130.
✅ Passed for params: limit=120, before=None, after=<message #20>, around=None, oldest_first=True. Expected 120, got 120.
✅ Passed for params: limit=50, before=None, after=<message #20>, around=None, oldest_first=True. Expected 50, got 50.
❌ Failed for params: limit=None, before=<message #130>, after=<message #20>, around=None, oldest_first=True. Expected 109, got 130.
✅ Passed for params: limit=75, before=<message #130>, after=<message #20>, around=None, oldest_first=True. Expected 75, got 75.
✅ Passed for params: limit=100, before=None, after=None, around=<message #75>, oldest_first=False. Expected around_test = lambda x: 100 <= x and x <= 101, got 100.
✅ Passed for params: limit=200, before=<message #130>, after=None, around=None, oldest_first=True. Expected 129, got 129.
✅ Passed for params: limit=200, before=None, after=<message #20>, around=None, oldest_first=False. Expected 130, got 130.
✅ Passed for params: limit=100, before=None, after=None, around=<message #75>, oldest_first=True. Expected around_test = lambda x: 100 <= x and x <= 101, got 100.
Successfully completed 12/15 tests.
```

**EmmmaTech/fix/async-iterators**
```
❌ Failed for params: limit=None, before=None, after=None, around=None, oldest_first=False. Expected 150, got 100.
✅ Passed for params: limit=50, before=None, after=None, around=None, oldest_first=False. Expected 50, got 50.
❌ Failed for params: limit=120, before=None, after=None, around=None, oldest_first=False. Expected 120, got 100.
✅ Passed for params: limit=None, before=<message #130>, after=None, around=None, oldest_first=False. Expected 129, got 129.
✅ Passed for params: limit=150, before=<message #130>, after=None, around=None, oldest_first=False. Expected 129, got 129.
✅ Passed for params: limit=80, before=<message #130>, after=None, around=None, oldest_first=False. Expected 80, got 80.
✅ Passed for params: limit=None, before=None, after=<message #20>, around=None, oldest_first=True. Expected 130, got 130.
✅ Passed for params: limit=120, before=None, after=<message #20>, around=None, oldest_first=True. Expected 120, got 120.
✅ Passed for params: limit=50, before=None, after=<message #20>, around=None, oldest_first=True. Expected 50, got 50.
✅ Passed for params: limit=None, before=<message #130>, after=<message #20>, around=None, oldest_first=True. Expected 109, got 109.
✅ Passed for params: limit=75, before=<message #130>, after=<message #20>, around=None, oldest_first=True. Expected 75, got 75.
✅ Passed for params: limit=100, before=None, after=None, around=<message #75>, oldest_first=False. Expected around_test = lambda x: 100 <= x and x <= 101, got 100.
✅ Passed for params: limit=200, before=<message #130>, after=None, around=None, oldest_first=True. Expected 129, got 129.
❌ Failed for params: limit=200, before=None, after=<message #20>, around=None, oldest_first=False. Expected 130, got 100.
✅ Passed for params: limit=100, before=None, after=None, around=<message #75>, oldest_first=True. Expected around_test = lambda x: 100 <= x and x <= 101, got 100.
Successfully completed 12/15 tests.
```

teaishealthy commented:
oops accidentally closed this

EmmmaTech commented:
could you share the testing code you used so I can locally test with my own changes? :>

teaishealthy commented:
@EmmmaTech 
https://gist.github.com/teaishealthy/7d46b6c0dfe8bf45e43208f4489fbaaf

make sure to configure the constants at the top of the file, then just call run_history_tests with a TextChannel

teaishealthy committed:
fix: resolve pagination limit cap for recent message history

teaishealthy reviewed (changes requested):
fixed the limit pagination issue.

**but:** defaulting `oldest_first=True` changes the default iteration order of `channel.history(limit=10)` from the newest to the oldest messages. this also creates counterintuitive behavior when using before, as it now walks forward (oldest first) instead of backward from the anchor.
